### PR TITLE
add JFT_CombinedRaw.py

### LINKED ISSUE #1: add code for JFT plotting of combined RAW files

### PR DESCRIPTION
**Pull request by HaochenWang1243**

### CODE REVIEW

**Comment by ntlhui** [comment]
Please add a description of what functionality this PR provides.

**Comment by ntlhui** [comment]
Please also make sure this change does not yield any `pylint` errors.

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                path=os.path.join(root,file)
+                print(f'reading {path}...')
+                t_signal_raw=generate_real_signal(path)
+                data_collection.append(t_signal_raw)
+    return data_collection
+
+
+
+if __name__ == '__main__':
+    t_signal_raw=read_raw_files('/home/haw057/RCT/RCT_Haochen/generate_signal_4testing/2017.07.06.RCT_Test/2017.07.06.RCT_Test/RUN_000001',1) #limit num of files read
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:108]: Please take in the input file as command line arguments.  Use https://docs.python.org/3/library/argparse.html to parse.  Store paths at `pathlib.Path` objects.

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        with open(dataFilePath, 'rb') as dataFile:
+            for i in range(nSamples):
+                sampleBytes = dataFile.read(4)
+                re, im = struct.unpack("<2h", sampleBytes)
+                signal_raw[i] = float(re) / 0x7fff + float(im) * 1j / 0x7fff
+        t_raw = np.arange(0, nSamples / f_s, 1/f_s)
+        print("Real Signal Generated.")
+        return t_raw, signal_raw
+    
+def read_raw_files(RAW_DATA_path,countLimit):
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:92]: Missing type hints.  Paths should use `pathlib.Path`.

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    data_collection=[]
+    count=0
+    for root, dirs, files in os.walk(RAW_DATA_path):
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:95]: use `pathlib.Path.iterdir` or `pathlib.Path.glob` or `pathlib.Path.rglob`

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        for file in files:
+            if file.startswith('RAW_DATA_') and count<countLimit:
+                count+=1
+                path=os.path.join(root,file)
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:99]: use `pathlib.Path.joinpath`

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            for i in range(nSamples):
+                sampleBytes = dataFile.read(4)
+                re, im = struct.unpack("<2h", sampleBytes)
+                signal_raw[i] = float(re) / 0x7fff + float(im) * 1j / 0x7fff
+        t_raw = np.arange(0, nSamples / f_s, 1/f_s)
+        print("Real Signal Generated.")
+        return t_raw, signal_raw
+    
+def read_raw_files(RAW_DATA_path,countLimit):
+    data_collection=[]
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:93]: use type hints

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py (hunk 2 above)
Comment (ntlhui) [reply DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:92]: Please add docstrings

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        for k in range(num_cycle):
+            signal[int(k*calm_period_j[i]+k*burst_period_j[i]):int((k+1)*(calm_period_j[i])+k*(burst_period_j[i]))]=0
+        # time shift
+        signal=np.roll(signal, int(random_offset_period_j[i]))
+        signal[:int(random_offset_period_j[i])]=0 
+        # add amplitude and noise
+        combined_signal+=amplitude_arr[i]*signal+noise_i[i]
+    return combined_signal
+
+def generate_real_signal(dataFilePath): #rawdata read and transform to signal
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:80]: Add docstring

Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:80]: use pathlib.Path and type hints

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            signal[int(k*calm_period_j[i]+k*burst_period_j[i]):int((k+1)*(calm_period_j[i])+k*(burst_period_j[i]))]=0
+        # time shift
+        signal=np.roll(signal, int(random_offset_period_j[i]))
+        signal[:int(random_offset_period_j[i])]=0 
+        # add amplitude and noise
+        combined_signal+=amplitude_arr[i]*signal+noise_i[i]
+    return combined_signal
+
+def generate_real_signal(dataFilePath): #rawdata read and transform to signal 
+        nSamples = int(os.path.getsize(dataFilePath) / 4)
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:81]: use pathlib.Path.stat

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        signal_raw = np.zeros(nSamples, dtype=np.complex128)
+        with open(dataFilePath, 'rb') as dataFile:
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:83]: missing encoding

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            for i in range(nSamples):
+                sampleBytes = dataFile.read(4)
+                re, im = struct.unpack("<2h", sampleBytes)
+                signal_raw[i] = float(re) / 0x7fff + float(im) * 1j / 0x7fff
+        t_raw = np.arange(0, nSamples / f_s, 1/f_s)
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:88]: inconsistent usage of spaces around operators

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    # Set axes label
+    ax.set_xlabel('time(s)', labelpad=20)
+    ax.set_ylabel('freq(MHz)', labelpad=20)
+    ax.set_zlabel('power', labelpad=20)
+    # ax.set_zlim(0,0.008)
+    plt.savefig('JFTplot_CombinedRawFiles_{}.jpg'.format(datetime.datetime.now()))
+    plt.show()
+    return t,freq,waterfall
+
+def make_signal(t_signal_raw):
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:51]: no docstring

Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:51]: add type hings

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        # 0 out the inactive periods
+        num_cycle=int(np.ceil(len(t)/(burst_period_j[i]+calm_period_j[i])))
+        for k in range(num_cycle):
+            signal[int(k*calm_period_j[i]+k*burst_period_j[i]):int((k+1)*(calm_period_j[i])+k*(burst_period_j[i]))]=0
+        # time shift
+        signal=np.roll(signal, int(random_offset_period_j[i]))
+        signal[:int(random_offset_period_j[i])]=0 
+        # add amplitude and noise
+        combined_signal+=amplitude_arr[i]*signal+noise_i[i]
+    return combined_signal
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:78]: Why is this in this file?  Maybe this should be in a separate file

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+from pathlib import Path
+
+f_s=10**6 #hyperparams
+f_c=172*10**6
+burst_period=20*10**3
+calm_period=1*10**6#40*10**4 #2x burst for demo purposes #1*10**6 #real value
+mean_amp=np.sqrt(2) #units 
+mean_val_noise=np.sqrt(2*np.pi)
+
+def generate_visualization(samples): #raw viz by latest method (june 2023)
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:18]: add docstring and type hints

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    seq_samples = np.array(samples)
+    arr_samples = np.reshape(seq_samples[0:FFT_LEN*int(len(samples) / FFT_LEN)], (FFT_LEN, int(len(samples) / FFT_LEN)), order='F')
+    f_samp = np.fft.fft(arr_samples, axis=0) / FFT_LEN
+    waterfall = np.power(np.abs(np.fft.fftshift(f_samp, axes=0)), 2)
+
+    # waterfall_extents = ((f_c - f_s / 2) / 1e6, (f_c + f_s / 2) / 1e6, len(seq_samples) / f_s, 0)
+
+    # freq_isolated_power = waterfall[fft_bin,:]
+    freq=np.arange(-f_s, f_s, 2*f_s/FFT_LEN)
+    t = np.arange(int(len(seq_samples)/FFT_LEN)) * FFT_LEN / f_s
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:32]: Inconsistent use of spaces around `=`

File: DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    fig.colorbar(surf, shrink=0.5, aspect=8)
+
+    # Set axes label
+    ax.set_xlabel('time(s)', labelpad=20)
+    ax.set_ylabel('freq(MHz)', labelpad=20)
+    ax.set_zlabel('power', labelpad=20)
+    # ax.set_zlim(0,0.008)
+    plt.savefig('JFTplot_CombinedRawFiles_{}.jpg'.format(datetime.datetime.now()))
+    plt.show()
+    return t,freq,waterfall
```
Comment (ntlhui) [inline DSP-PING-main/dsp-summer2023/JFT_CombinedRaw.py:49]: Possible MVC violation.  Is this code meant to be reusable?

**Review by ntlhui — CHANGES_REQUESTED** [review]

**Comment by codersupreme99101** [comment]
closed for the code to be PR'ed into the multi repo

**Pull request closed without merging**